pkg/pattern/drain/chunk.go

```diff
@@ -66,11 +66,37 @@ func (c Chunk) ForRange(start, end model.Time) []logproto.PatternSample {
 			return c.Samples[i].Timestamp >= end
 		})
 	}
-	return c.Samples[lo:hi]
+	if step == timeResolution {
+		return c.Samples[lo:hi]
+	}
+
+	// Re-scale samples into step-sized buckets
+	currentStep := truncateTimestamp(c.Samples[lo].Timestamp, step)
+	aggregatedSamples := []logproto.PatternSample{
```

Comment:
nit: if you can figure in advance the amount of sample, then allocating the right amount of sample will be faster. Even if it's off by 1, growing will always perform poorly.